src/input/joystick.c

```diff
@@ -0,0 +1,714 @@
+            input->value = 0;
+            return 0;
+        }
+        // Element precedence. When two joystick elements are required for the input
+        // to be registered, the value is obtained from the highest priority element.
+        // Element priority in descending order: axis, trackball, button, hat
+        if (input->element == JOYSTICK_ELEMENT_NONE || current_element < input->element) {
+            input->element = current_element;
+            input->value = current_value;
+        }
+    }
+    return input->value != 0;
+}
+
+static void set_input_state(mapped_input *input)
+{
+    if (input->value) {
+        if (input->state == INPUT_STATE_IS_UP || input->state == INPUT_STATE_WENT_UP) {
+            input->state = INPUT_STATE_WENT_DOWN;
+        } else {
+            input->state = INPUT_STATE_IS_DOWN;
+        }
+    } else {
+        if (input->state == INPUT_STATE_IS_DOWN || input->state == INPUT_STATE_WENT_DOWN) {
+            input->state = INPUT_STATE_WENT_UP;
+        } else {
+            input->state = INPUT_STATE_IS_UP;
+        }
+    }
+}
+
+static int get_joystick_input_for_action(mapping_action action, mapped_input *input)
+{
+    static mapped_input dummy_input;
+    if (!input) {
+        input = &dummy_input;
+    }
+
+    input->value = 0;
+    input->element = JOYSTICK_ELEMENT_NONE;
+
+    for (int i = 0; i < JOYSTICK_MAX_CONTROLLERS; ++i) {
+        const joystick_info *joystick = &data.joystick[i];
+        if (!joystick->connected) {
+            continue;
+        }
+        if(get_input_for_mapping(joystick, &joystick->model->first_mapping[action], input) ||
+           get_input_for_mapping(joystick, &joystick->model->second_mapping[action], input)) {
+            set_input_state(input);
+            return 1;
+        }
+    }
+    set_input_state(input);
+    return 0;
+}
+
+static void translate_input_for_element(mapped_input *input, joystick_element translated_element)
+{
+    if (input->element == translated_element) {
+        return;
+    }
+    if (translated_element == JOYSTICK_ELEMENT_AXIS) {
+        switch(input->element)
+        {
+            case JOYSTICK_ELEMENT_TRACKBALL:
+                input->value *= 5; // TODO is this a high enough value? I doubt it
+                break;
+            default:
+                input->value *= 32767;
+        }
+    } else if (translated_element == JOYSTICK_ELEMENT_TRACKBALL) {
+        switch (input->element) {
+            case JOYSTICK_ELEMENT_AXIS:
+                input->value /= 5; // TODO is this a high enough value? I doubt it
+                break;
+            default:
+                input->value *= 5;
+        }
+    } else if (translated_element == JOYSTICK_ELEMENT_BUTTON || translated_element == JOYSTICK_ELEMENT_HAT) {
+        input->value = (input->value != 0) ? 1 : 0;
+    }
+    input->element = translated_element;
+}
+
+static int rescale_axis(mapped_input *inputs)
+{
+    //radial and scaled dead_zone
```

Comment:
One space after `//` please. Also format like normal sentences, with capital letter at the start.